lib/resumption/handover.go

```diff
@@ -0,0 +1,112 @@
+// Teleport
+// Copyright (C) 2024 Gravitational, Inc.
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+package resumption
+
+import (
+	"context"
+	"io"
+	"net"
+	"net/netip"
+	"time"
+
+	"github.com/gravitational/trace"
+
+	"github.com/gravitational/teleport/lib/multiplexer"
+	"github.com/gravitational/teleport/lib/utils"
+)
+
+func (r *SSHServerWrapper) attemptHandover(conn *multiplexer.Conn, token resumptionToken) {
+	handoverConn, err := r.dialHandover(token)
+	if err != nil {
+		if trace.IsNotFound(err) {
+			r.log.Debug("Resumable connection not found or already deleted.")
+			_, _ = conn.Write([]byte{notFoundServerExchangeTag})
+			return
+		}
+		r.log.WithError(err).Error("Error while connecting to handover socket.")
+		return
+	}
+	defer handoverConn.Close()
+
+	var remoteIP netip.Addr
+	if t, _ := conn.RemoteAddr().(*net.TCPAddr); t != nil {
+		remoteIP, _ = netip.AddrFromSlice(t.IP)
+	}
+	remoteIP16 := remoteIP.As16()
+
+	if _, err := handoverConn.Write(remoteIP16[:]); err != nil {
+		if !utils.IsOKNetworkError(err) {
+			r.log.WithError(err).Error("Error while forwarding remote address to handover socket.")
+		}
+		return
+	}
+
+	r.log.Debug("Forwarding resuming connection to handover socket.")
+	_ = utils.ProxyConn(context.Background(), conn, handoverConn)
+}
+
+func (r *SSHServerWrapper) setupHandoverListener(ctx context.Context, token resumptionToken, entry *connEntry) error {
```

Comment:
Shouldn't this also be in `handover_unix.go`? The current organization is a bit difficult to parse, is there a cleaner way to separate what's `_unix.go` and what's `_other.go`?

Reply:
I get the general idea, but it's still not quite clear to me what the dividing line is. For example it seems quite clear to me that `setupHandoverListener` (now `startHandoverListener`) does indeed "need Unix domain sockets", because it listens to a Unix domain socket.

It's also causing me to strain to walk through the logic in the non-UNIX case, e.g. we need a check like

https://github.com/gravitational/teleport/blob/aefbc4574ca6c1c174e6c9b36428587d66e8bfa3/lib/resumption/server_exchange.go#L120

with the `&& !trace.IsNotImplemented(err)` which I believe is to account for the error returned here (in `_other.go`)

https://github.com/gravitational/teleport/blob/aefbc4574ca6c1c174e6c9b36428587d66e8bfa3/lib/resumption/handover_other.go#L28-L30

but which is called by `startHandoverListener` which is just in `handover.go`

https://github.com/gravitational/teleport/blob/aefbc4574ca6c1c174e6c9b36428587d66e8bfa3/lib/resumption/handover.go#L62-L72

It looks like this is the only instance of this now, but as this section evolves over time it seems liable to become difficult to reason about and therefore maintain. 

IMO a more easily comprehensible approach would be to try and cordon off these handover parts into their own `HandoverClient` and `HandoverServer` structs with clear public APIs, which can be turned into dummy APIs in the `_other.go` (you can look to the `_nop.go` family of files for prior art).

Reply:
I've just gotten rid of the build tag split because I realized that modern-ish Windows actually supports UNIX sockets anyway - so the single implementation with sockets in a directory wouldn't be the blocker for some eventual Windows support of `ssh_service` anyway.